src/optimizer/expression_heuristics.cpp

```diff
@@ -0,0 +1,233 @@
+}
+
+void ExpressionHeuristics::ReorderExpressions(vector<unique_ptr<Expression>> &expressions) {
+
+	struct expr {
+		index_t idx;
+		index_t cost;
+
+		bool operator==(const expr &p) const {
+		return idx == p.idx && cost == p.cost;
+		}
+		bool operator<(const expr &p) const {
+		return cost < p.cost || (cost == p.cost && idx < p.idx);
+		}
+	};
+
+	vector<expr> exprCosts;
+	//iterate expressions, get cost for each one and order by that cost
+	for (index_t i = 0; i < expressions.size(); i++) {
+		exprCosts.push_back({i, Cost(*expressions[i])});
+	}
+
+	sort(exprCosts.begin(), exprCosts.end());
+	assert(exprCosts.size() == expressions.size());
+
+	//for all elements to put in place
+	for(index_t i = 0; i < expressions.size() - 1; ++i) {
+		//while the element i is not yet in place
+		while(i != exprCosts[i].idx) {
+			//swap it with the element at its final place
+			index_t alt = exprCosts[i].idx;
+			swap(expressions[i], expressions[alt]);
+			swap(exprCosts[i], exprCosts[alt]);
+		}
+	}
+}
+
+index_t ExpressionHeuristics::Cost(Expression &expr) {
+	switch (expr.expression_class) {
+		case ExpressionClass::BOUND_CASE: {
+			auto &case_expr = (BoundCaseExpression &)expr;
+
+			//CASE WHEN check THEN result_if_true ELSE result_if_false END
+			return Cost(*case_expr.check) + Cost(*case_expr.result_if_true) + Cost(*case_expr.result_if_false) + 5;
+		}
+		case ExpressionClass::BOUND_CAST: {
+			//OPERATOR_CAST
+			auto &cast_expr = (BoundCastExpression &)expr;
+
+			//determine cast cost by comparing cast_expr.source_type and cast_expr_target_type
+			index_t cast_cost = 0;
+			if (cast_expr.target_type != cast_expr.source_type) {
+				//if cast from or to varchar
+				//TODO: we might want to add more cases
+				if (cast_expr.target_type == SQLType::VARCHAR || cast_expr.source_type == SQLType::VARCHAR) {
+					cast_cost = 200;
+				} else {
+					cast_cost = 5;
+				}
+			}
+
+			return Cost(*cast_expr.child) + cast_cost;
+		}
+		case ExpressionClass::BOUND_COMPARISON: {
+			auto &comp_expr = (BoundComparisonExpression &)expr;
+
+			//COMPARE_EQUAL, COMPARE_NOTEQUAL, COMPARE_GREATERTHAN, COMPARE_GREATERTHANOREQUALTO, COMPARE_LESSTHAN, COMPARE_LESSTHANOREQUALTO
+			return Cost(*comp_expr.left) + 5 + Cost(*comp_expr.right);
+		}
+		case ExpressionClass::BOUND_CONJUNCTION: {
+			auto &conj_expr = (BoundConjunctionExpression &)expr;
+
+			//CONJUNCTION_AND, CONJUNCTION_OR
+			return Cost(*conj_expr.left) + 5 + Cost(*conj_expr.right);
+		}
+		case ExpressionClass::BOUND_FUNCTION: {
+			auto &func_expr = (BoundFunctionExpression &)expr;
+
+			index_t cost_children = 0;
+			for (auto &child : func_expr.children) {
+				cost_children += Cost(*child);
+			}
+
+			auto func_name = func_expr.function.name;
+
+			//ADD(+), SUB(-), BITWISE AND(&), BITWISE OR(#), RIGHTSHIFT(>>), LEFTSHIFT(<<), abs()
+			if (func_name == "+" || func_name == "-" || func_name == "&" || func_name == "#" || func_name == ">>" || func_name == "<<" || func_name == "abs") {
```

Comment:
Please replace this with a "unordered_map<string, index_t> costs" and add the function names to that map, then perform a lookup on that map. This will be more efficient and cleaner.